lib/tpm_policy.c

```diff
@@ -0,0 +1,254 @@
+            total_indices_for_this_alg+=(((group_val>>3)&1)+((group_val>>2)&1)+((group_val>>1)&1)+(group_val&1));
+        }
+        expected_pcr_input_file_size += (total_indices_for_this_alg * 
+            get_size_from_alg(pctx->pcr_policy_options.pcr_selections.pcrSelections[i].hash));
+        /*
+         * indexing with k is used as valid map while populating from input file if 
+         * file_size_test passes
+         */
+        pcr_values->count += total_indices_for_this_alg;
+        for (k=0; k<total_indices_for_this_alg; k++) {
+            pcr_values->digests[dgst_cnt+k].t.size = 
+                get_size_from_alg(pctx->pcr_policy_options.pcr_selections.pcrSelections[i].hash);
+        }
+        dgst_cnt++;
+
+        total_indices_for_this_alg=0;
+    }
+
+    int file_size_test = (pcr_values->count == 0) ? 1:0;
+    if (file_size_test && pctx->pcr_policy_options.is_raw_pcrs_file) {
+        long filesize = 0;
+        bool result = files_get_file_size(pctx->pcr_policy_options.raw_pcrs_file, &filesize);
+        if (!result) {
+                return false;
+        } else {
+            if (filesize != expected_pcr_input_file_size) {
+                LOG_ERR("pcr-input-file filesize does not match pcr set-list");
+                return false;
+            } 
+        }
+    } 
+    return true;
+}
+
+TPM_RC build_pcr_policy(CREATE_POLICY_CTX *pctx) {
+    //PCR inputs validation
+    if (pctx->pcr_policy_options.is_set_list == false) {
+        LOG_ERR("Need the pcr list to account for in the policy.");
+        return TPM_RC_NO_RESULT;
+    }
+
+    // Calculate digest( with authhash alg) of pcrvalues in variable pcr_digest
+    TPM_RC rval=0;
+    TPM2B_DIGEST pcr_digest = TPM2B_TYPE_INIT(TPM2B_DIGEST, buffer);
+    TPML_DIGEST pcr_values;
+
+    bool result = evaluate_populate_pcr_digests(pctx, &pcr_values);
+    if (!result) {
+        return TPM_RC_NO_RESULT;
+    } 
+
+    if (pctx->pcr_policy_options.is_raw_pcrs_file) {
+        FILE *fp = fopen (pctx->pcr_policy_options.raw_pcrs_file, "rb");
+        if (fp == NULL) {
+            LOG_ERR("Cannot open pcr-input-file %s", pctx->pcr_policy_options.raw_pcrs_file);
+            return TPM_RC_NO_RESULT;
+        }
+       // Bank hashAlg values dictates the order of the list of digests
+        unsigned i;
+        for(i=0; i<pcr_values.count; i++) {
+            size_t sz = fread(&pcr_values.digests[i].t.buffer, 1, pcr_values.digests[i].t.size, fp);
+            if (sz != pcr_values.digests[i].t.size) {
+                const char *msg = ferror(fp) ? strerror(errno) :
+                        "end of file reached";
+                LOG_ERR("Reading from file \"%s\" failed: %s",
+                        pctx->pcr_policy_options.raw_pcrs_file, msg);
+                fclose(fp);
+                return TPM_RC_NO_RESULT;
+            }
+        }
+        fclose(fp);
+    } else {
```

Comment:
Big blocks of if...else statements are hard to follow. Perhaps use static internal helper functions to abstract whats going on... this is much easier to read:

```
if (a) {
  do_something_a();
} else {
  do_something_b();
}
```

Versus very long blocks of code.